File: server/src/cve/cvelist-importer.ts
import { join, sep } from "path";
import { promises as fs } from "fs";
import simpleGit from "simple-git";
import { db, dbGet, dbAll, DATA_DIR } from "../db/database";
import { CvelistSyncState } from "../types";
import { checkEnrichmentAlerts } from "./enrichment-alerter";
import { evaluateCveTargetsFromCvelist } from "./cvelist-engine";

const REPO_URL = "https://github.com/CVEProject/cvelistV5.git";

export function getRepoDir(): string {
  return join(DATA_DIR, "cvelistV5");
}

interface CvelistSyncResult {
  count: number;
  repoVersion: string;
}

interface VersionRange {
  version?: string;
  lessThan?: string;
  lessThanOrEqual?: string;
  versionType?: string;
  status?: string;
}

interface AffectedEntry {
  vendor?: string;
  product?: string;
  versions?: VersionRange[];
  defaultStatus?: string;
}

interface CvelistRecord {
  cveMetadata?: {
    cveId?: string;
    state?: string;
    datePublished?: string;
    dateUpdated?: string;
  };
  containers?: {
    cna?: {
      title?: string;
      descriptions?: Array<{ lang: string; value: string }>;
      affected?: AffectedEntry[];
    };
  };
}

interface CvelistRow {
  cveId: string;
  state: string;
  cnaDescription: string | null;
  cnaTitle: string | null;
  datePublished: string | null;
  dateUpdated: string | null;
  affected: AffectedEntry[];
}

function extractCvelist(record: CvelistRecord): CvelistRow | null {
  const meta = record.cveMetadata;
  if (!meta?.cveId || !meta.state) return null;
  if (!meta.cveId.startsWith("CVE-")) return null;

  const cna = record.containers?.cna;
  const descriptions = cna?.descriptions ?? [];
  const enDesc = descriptions.find(
    (d) => d.lang === "en" || d.lang.startsWith("en")
  )?.value ?? null;

  return {
    cveId: meta.cveId,
    state: meta.state,
    cnaDescription: enDesc,
    cnaTitle: cna?.title ?? null,
    datePublished: meta.datePublished ?? null,
    dateUpdated: meta.dateUpdated ?? null,
    affected: cna?.affected ?? [],
  };
}

async function walkJsonFiles(dir: string): Promise<string[]> {
  const results: string[] = [];
  let entries: import("fs").Dirent[];
  try {
    entries = await fs.readdir(dir, { withFileTypes: true });
  } catch {
    return results;
  }
  for (const entry of entries) {
    const fullPath = join(dir, entry.name);
    if (entry.isDirectory()) {
      const nested = await walkJsonFiles(fullPath);
      results.push(...nested);
    } else if (entry.isFile() && entry.name.endsWith(".json") && entry.name.startsWith("CVE-")) {
      results.push(fullPath);
    }
  }
  return results;
}

async function getChangedFiles(repoDir: string, fromHash: string): Promise<string[] | null> {
  try {
    const result = await simpleGit(repoDir).raw([
      "diff", "--name-only", fromHash, "HEAD",
    ]);
    return result
      .trim()
      .split("\n")
      .filter((f) => f.endsWith(".json") && f.includes("CVE-"))
      .map((f) => join(repoDir, f.replace(/\//g, sep)));
  } catch {
    return null;
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let _upsertCve: any = null;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
let _upsertAffected: any = null;

function ensureStmts(): void {
  if (!_upsertCve) {
    _upsertCve = db.prepare(
      `INSERT OR REPLACE INTO cvelist_cves
         (cve_id, state, cna_description, cna_title, date_published, date_updated, synced_at)
       VALUES (?, ?, ?, ?, ?, ?, ?)`
    );
  }
  if (!_upsertAffected) {
    _upsertAffected = db.prepare(
      `INSERT INTO cvelist_affected (cve_id, vendor, product, versions_json, default_status)
       VALUES (?, ?, ?, ?, ?)`
    );
  }
}

export async function syncCvelist(): Promise<CvelistSyncResult> {
  const repoDir = getRepoDir();
  const git = simpleGit();

  // Clone (blobless: full history for diff, no blob content upfront) or pull
  try {
    await fs.access(join(repoDir, ".git"));
    console.log("[cvelist] pulling latest changes...");
    await simpleGit(repoDir).pull();
  } catch {
    console.log("[cvelist] cloning repository (blobless, may take a moment)...");
    await git.clone(REPO_URL, repoDir, ["--filter=blob:none", "--no-tags"]);
  }

  // Get new HEAD hash
  let repoVersion = "";
  try {
    const log = await simpleGit(repoDir).log({ maxCount: 1 });
    repoVersion = log.latest?.hash ?? "";
  } catch {
    // non-critical
  }

  // Get last processed hash from nvd_feed_state
  const stateRow = dbGet<{ sha256: string | null }>(
    "SELECT sha256 FROM nvd_feed_state WHERE feed_name = 'cvelist'"
  );
  const lastHash = stateRow?.sha256 ?? null;

  // Determine which files to process
  if (lastHash && repoVersion && lastHash === repoVersion) {
    console.log("[cvelist] already up to date, nothing to process");
    await checkEnrichmentAlerts();
    await evaluateCveTargetsFromCvelist();
    return { count: 0, repoVersion };
  }

  let filesToProcess: string[] | null = null;
  if (lastHash && repoVersion) {
    console.log(`[cvelist] computing diff from ${lastHash.slice(0, 8)} to ${repoVersion.slice(0, 8)}...`);
    filesToProcess = await getChangedFiles(repoDir, lastHash);
    if (filesToProcess !== null) {
      console.log(`[cvelist] incremental: ${filesToProcess.length} changed files`);
    } else {
      console.log("[cvelist] git diff failed, falling back to full walk");
    }
  }

  const jsonFiles = filesToProcess ?? await walkJsonFiles(join(repoDir, "cves"));
  if (!filesToProcess) {
    console.log(`[cvelist] full walk: ${jsonFiles.length} files`);
  }

  const syncedAt = new Date().toISOString();
  ensureStmts();

  let count = 0;
  const BATCH_SIZE = 500;
  let batch: CvelistRow[] = [];

  const flush = () => {
    if (batch.length === 0) return;
    db.exec("BEGIN IMMEDIATE");
    try {
      for (const row of batch) {
        _upsertCve.run(
          row.cveId, row.state, row.cnaDescription, row.cnaTitle,
          row.datePublished, row.dateUpdated, syncedAt
        );
        // Delete old affected entries and re-insert
        db.prepare("DELETE FROM cvelist_affected WHERE cve_id = ?").run(row.cveId);
        for (const aff of row.affected) {
          if (!aff.product) continue;
          _upsertAffected.run(
            row.cveId,
            aff.vendor?.toLowerCase() ?? null,
            aff.product.toLowerCase(),
            aff.versions ? JSON.stringify(aff.versions) : null,
            aff.defaultStatus ?? null
          );
        }
      }
      db.exec("COMMIT");
    } catch (err) {
      db.exec("ROLLBACK");
      throw err;
    }
    count += batch.length;
    batch = [];
  };

  const LOG_INTERVAL = 10_000;
  let filesRead = 0;

  for (const filePath of jsonFiles) {
    let record: CvelistRecord;
    try {
      const raw = await fs.readFile(filePath, "utf-8");
      record = JSON.parse(raw) as CvelistRecord;
    } catch {
      continue;
    }

    const row = extractCvelist(record);
    if (!row) continue;

    batch.push(row);
    if (batch.length >= BATCH_SIZE) flush();

    filesRead++;
    if (filesRead % LOG_INTERVAL === 0) {
      console.log(`[cvelist] processed ${filesRead.toLocaleString()} / ${jsonFiles.length.toLocaleString()} files (${count.toLocaleString()} upserted so far)...`);
    }
  }
  flush();

  console.log(`[cvelist] upserted ${count} CVE records`);

  // Update sync state
  db.prepare(
    `INSERT OR REPLACE INTO nvd_feed_state (feed_name, sha256, total_cves, imported_at)
     VALUES ('cvelist', ?, ?, ?)`
  ).run(repoVersion, count, syncedAt);

  // Run enrichment checks and target evaluation
  await checkEnrichmentAlerts();
  await evaluateCveTargetsFromCvelist();

  return { count, repoVersion };
}

export function getCvelistSyncState(): CvelistSyncState {
  const row = dbGet<{ total: number; rejected_count: number; last_synced_at: string | null }>(
    `SELECT COUNT(*) AS total,
            SUM(CASE WHEN state = 'REJECTED' THEN 1 ELSE 0 END) AS rejected_count,
            MAX(synced_at) AS last_synced_at
     FROM cvelist_cves`
  );
  const stateRow = dbGet<{ sha256: string | null; imported_at: string | null }>(
    "SELECT sha256, imported_at FROM nvd_feed_state WHERE feed_name = 'cvelist'"
  );
  return {
    total: row?.total ?? 0,
    rejected_count: row?.rejected_count ?? 0,
    last_synced_at: stateRow?.imported_at ?? null,
    is_syncing: false,
    last_repo_version: stateRow?.sha256 ? stateRow.sha256.slice(0, 8) : null,
  };
}

export type { AffectedEntry, VersionRange };
export { dbAll };

File: vigil/src/cve/cvelist_importer.rs
use std::{
    path::{Path, PathBuf},
    process::Command,
    sync::{Arc, Mutex},
};

use serde_json::Value;

use crate::{db::DbPool, types::SyncProgress};

const REPO_URL: &str = "https://github.com/CVEProject/cvelistV5.git";
const BATCH_SIZE: usize = 500;

pub struct CvelistSyncResult {
    pub count: usize,
    pub repo_version: String,
}

struct CvelistRow {
    cve_id: String,
    state: String,
    cna_description: Option<String>,
    cna_title: Option<String>,
    date_published: Option<String>,
    date_updated: Option<String>,
    affected: Vec<AffectedEntry>,
}

struct AffectedEntry {
    vendor: Option<String>,
    product: String,
    versions_json: Option<String>,
    default_status: Option<String>,
}

pub async fn sync_cvelist(
    db: &DbPool,
    data_dir: &str,
    progress: Arc<Mutex<Option<SyncProgress>>>,
) -> anyhow::Result<CvelistSyncResult> {
    let repo_dir = PathBuf::from(data_dir).join("cvelistV5");
    let repo_dir_str = repo_dir.to_string_lossy().to_string();

    let db = db.clone();
    tokio::task::spawn_blocking(move || -> anyhow::Result<CvelistSyncResult> {
        let set_progress = |stage: &str, message: String, files_done: usize, files_total: usize| {
            if let Ok(mut p) = progress.lock() {
                *p = Some(SyncProgress { stage: stage.to_string(), message, files_done, files_total });
            }
        };

        // Clone or pull
        let is_new_clone = !repo_dir.join(".git").exists();
        if is_new_clone {
            tracing::info!("[cvelist] cloning repository (blobless, may take a moment)");
            set_progress("git_clone", "Cloning CVE Program repository (blobless)…".to_string(), 0, 0);
            git_clone(REPO_URL, &repo_dir_str, &["--filter=blob:none", "--no-tags"])?;
        } else {
            tracing::info!("[cvelist] pulling latest changes");
            set_progress("git_pull", "Pulling latest CVE list changes…".to_string(), 0, 0);
            git_pull(&repo_dir_str)?;
        }

        let repo_version = git_head(&repo_dir_str).unwrap_or_default();

        // Check if already up to date
        let last_hash: Option<String> = {
            let conn = db.lock().unwrap();
            conn.query_row(
                "SELECT sha256 FROM nvd_feed_state WHERE feed_name = 'cvelist'",
                [],
                |row| row.get(0),
            )
            .ok()
        };

        if let Some(ref last) = last_hash {
            if !last.is_empty() && *last == repo_version {
                tracing::info!("[cvelist] already up to date, nothing to process");
                return Ok(CvelistSyncResult { count: 0, repo_version });
            }
        }

        // Determine files to process
        let files_to_process: Vec<PathBuf> = if let Some(ref last) = last_hash {
            if !last.is_empty() && !repo_version.is_empty() {
                match git_diff_files(&repo_dir_str, last) {
                    Ok(changed) => {
                        tracing::info!(count = changed.len(), "[cvelist] incremental: changed files");
                        changed
                            .into_iter()
                            .map(|f| repo_dir.join(f))
                            .filter(|p| {
                                p.extension().and_then(|e| e.to_str()) == Some("json")
                                    && p.file_name()
                                        .and_then(|n| n.to_str())
                                        .map(|n| n.starts_with("CVE-"))
                                        .unwrap_or(false)
                            })
                            .collect()
                    }
                    Err(_) => {
                        tracing::warn!("[cvelist] git diff failed, falling back to full walk");
                        walk_json_files(&repo_dir.join("cves"))
                    }
                }
            } else {
                walk_json_files(&repo_dir.join("cves"))
            }
        } else {
            walk_json_files(&repo_dir.join("cves"))
        };

        let total_files = files_to_process.len();
        tracing::info!(count = total_files, "[cvelist] files to process");
        set_progress("parse", format!("Parsing {} CVE files…", total_files), 0, total_files);

        // Sequential parse with progress every 5 000 files
        let mut parsed: Vec<CvelistRow> = Vec::with_capacity(total_files);
        for (i, path) in files_to_process.iter().enumerate() {
            if i > 0 && i % 5_000 == 0 {
                set_progress("parse", format!("Parsing CVE files… {}/{}", i, total_files), i, total_files);
            }
            let Some(raw) = std::fs::read_to_string(path).ok() else { continue };
            let Ok(record) = serde_json::from_str::<Value>(&raw) else { continue };
            if let Some(row) = extract_cvelist(&record) {
                parsed.push(row);
            }
        }

        let total_parsed = parsed.len();
        tracing::info!(count = total_parsed, "[cvelist] parsed CVE records — starting upsert");
        set_progress("insert", format!("Inserting {} CVE records…", total_parsed), 0, total_parsed);

        let synced_at = chrono::Utc::now().to_rfc3339();
        let mut count = 0usize;

        for chunk in parsed.chunks(BATCH_SIZE) {
            let conn = db.lock().unwrap();
            conn.execute_batch("BEGIN IMMEDIATE")?;
            let result: rusqlite::Result<()> = (|| {
                let mut stmt_cve = conn.prepare_cached(
                    "INSERT OR REPLACE INTO cvelist_cves
                     (cve_id, state, cna_description, cna_title, date_published, date_updated, synced_at)
                     VALUES (?1,?2,?3,?4,?5,?6,?7)",
                )?;
                let mut stmt_del = conn.prepare_cached(
                    "DELETE FROM cvelist_affected WHERE cve_id = ?1",
                )?;
                let mut stmt_aff = conn.prepare_cached(
                    "INSERT INTO cvelist_affected
                     (cve_id, vendor, product, versions_json, default_status)
                     VALUES (?1,?2,?3,?4,?5)",
                )?;
                for row in chunk {
                    stmt_cve.execute(rusqlite::params![
                        row.cve_id, row.state, row.cna_description, row.cna_title,
                        row.date_published, row.date_updated, synced_at,
                    ])?;
                    stmt_del.execute(rusqlite::params![row.cve_id])?;
                    for aff in &row.affected {
                        stmt_aff.execute(rusqlite::params![
                            row.cve_id, aff.vendor, aff.product,
                            aff.versions_json, aff.default_status,
                        ])?;
                    }
                }
                Ok(())
            })();
            match result {
                Ok(()) => conn.execute_batch("COMMIT")?,
                Err(e) => { conn.execute_batch("ROLLBACK")?; return Err(e.into()); }
            }
            count += chunk.len();
            // Emit insert progress every 10 batches (5 000 records)
            if (count / BATCH_SIZE) % 10 == 0 {
                set_progress("insert", format!("Inserting CVE records… {}/{}", count, total_parsed), count, total_parsed);
            }
        }

        // Update sync state
        {
            let conn = db.lock().unwrap();
            conn.execute(
                "INSERT OR REPLACE INTO nvd_feed_state
                 (feed_name, sha256, total_cves, imported_at)
                 VALUES ('cvelist', ?1, ?2, ?3)",
                rusqlite::params![repo_version, count as i64, synced_at],
            )?;
        }

        tracing::info!(count, "[cvelist] upserted CVE records");
        Ok(CvelistSyncResult { count, repo_version })
    })
    .await?
}

fn extract_cvelist(record: &Value) -> Option<CvelistRow> {
    let meta = &record["cveMetadata"];
    let cve_id = meta["cveId"].as_str()?.to_string();
    let state = meta["state"].as_str()?.to_string();
    if !cve_id.starts_with("CVE-") { return None; }

    let cna = &record["containers"]["cna"];
    let descriptions = cna["descriptions"].as_array();
    let cna_description = descriptions.and_then(|ds| {
        ds.iter()
            .find(|d| d["lang"].as_str().map(|l| l == "en" || l.starts_with("en")).unwrap_or(false))
            .and_then(|d| d["value"].as_str())
            .map(|s| s.to_string())
    });

    let cna_title = cna["title"].as_str().map(|s| s.to_string());
    let date_published = meta["datePublished"].as_str().map(|s| s.to_string());
    let date_updated = meta["dateUpdated"].as_str().map(|s| s.to_string());

    let affected = cna["affected"]
        .as_array()
        .map(|arr| {
            arr.iter()
                .filter_map(|a| {
                    let product = a["product"].as_str()?.to_string();
                    let vendor = a["vendor"].as_str().map(|v| v.to_lowercase());
                    let versions_json = a.get("versions")
                        .and_then(|v| if v.is_array() { Some(v.to_string()) } else { None });
                    let default_status = a["defaultStatus"].as_str().map(|s| s.to_string());
                    Some(AffectedEntry { vendor, product: product.to_lowercase(), versions_json, default_status })
                })
                .collect()
        })
        .unwrap_or_default();

    Some(CvelistRow { cve_id, state, cna_description, cna_title, date_published, date_updated, affected })
}

fn walk_json_files(dir: &Path) -> Vec<PathBuf> {
    let mut results = Vec::new();
    let Ok(entries) = std::fs::read_dir(dir) else { return results };
    for entry in entries.flatten() {
        let path = entry.path();
        if path.is_dir() {
            results.extend(walk_json_files(&path));
        } else if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
            if name.ends_with(".json") && name.starts_with("CVE-") {
                results.push(path);
            }
        }
    }
    results
}

fn git_clone(url: &str, target: &str, extra_args: &[&str]) -> anyhow::Result<()> {
    let status = Command::new("git")
        .arg("clone").args(extra_args).arg(url).arg(target)
        .env("GIT_TERMINAL_PROMPT", "0")
        .status()?;
    if !status.success() { anyhow::bail!("git clone failed for {}", url); }
    Ok(())
}

fn git_pull(repo_dir: &str) -> anyhow::Result<()> {
    let status = Command::new("git")
        .args(["pull"]).current_dir(repo_dir)
        .env("GIT_TERMINAL_PROMPT", "0")
        .status()?;
    if !status.success() { anyhow::bail!("git pull failed in {}", repo_dir); }
    Ok(())
}

fn git_head(repo_dir: &str) -> anyhow::Result<String> {
    let output = Command::new("git")
        .args(["rev-parse", "HEAD"]).current_dir(repo_dir)
        .env("GIT_TERMINAL_PROMPT", "0")
        .output()?;
    Ok(String::from_utf8(output.stdout)?.trim().to_string())
}

fn git_diff_files(repo_dir: &str, from_hash: &str) -> anyhow::Result<Vec<String>> {
    let output = Command::new("git")
        .args(["diff", "--name-only", from_hash, "HEAD"]).current_dir(repo_dir)
        .env("GIT_TERMINAL_PROMPT", "0")
        .output()?;
    let lines = String::from_utf8(output.stdout)?;
    Ok(lines.lines()
        .filter(|f| f.ends_with(".json") && f.contains("CVE-"))
        .map(|f| f.to_string())
        .collect())
}
